--- parser_common_ical.module	2008-03-08 11:33:12.000000000 -0500
+++ parser_common_ical.module	2008-05-30 15:48:17.781361600 -0400
@@ -9,8 +9,8 @@
 /**
  * Implementation of hook_help().
  */
-function parser_common_ical_help($section) {
-  switch($section) {
+function parser_common_ical_help($path,$arg) {
+  switch($path) {
     case 'admin/modules#description':
       return t('Provide a common ical parser for FeedAPI-compatible modules.');
     case 'feedapi/full_name':
@@ -120,7 +119,9 @@ function _parser_common_ical_parse($feed
 //    $item->options->teaser = $teaser;
     $item->options->location = $event['URL'];
     $item->options->original_author = $original_author;
-    $item->options->timestamp = date_format(date_ical_date($event['CREATED']),'U'); // TODO tz?
+    if($event['CREATED']){
+      $item->options->timestamp = date_format(date_ical_date($event['CREATED']),'U'); // TODO tz?
+    }
     $item->options->original_url = $original_url;
     $item->options->guid = $event['UID']; // intention
     $item->options->tags = array();// check
